### PR TITLE
HEXDEV-732: GLRM Memory Allocation Reduction

### PR DESCRIPTION
**Pull request by wendycwong**
This PR aims to solve this issue: https://0xdata.atlassian.net/projects/HEXDEV/issues/HEXDEV-732?filter=allopenissues .

There are a lot of array allocation at various places in the algo, I plan to get rid of them.

### CODE REVIEW

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java [hunk 1]
```diff
@@ -151,7 +174,9 @@ public GlrmMojoModel(String[] columns, String[][] domains, String responseColumn
    * @return
    */
   public double applyBestAlpha(double[] u, double[] x, double[] grad, double[] a, double oldObj, Random random) {
-    double[] bestX = new double[x.length];
+    //double[] bestX = new double[x.length];
```
Comment (Pscheidl) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: Remove this ?

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java [hunk 2]
```diff
@@ -151,7 +174,9 @@ public GlrmMojoModel(String[] columns, String[][] domains, String responseColumn
...
+    double[] bestX = _tbest;
```
Comment (Pscheidl) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: The original code creates new array which `bestX` variable points to. This piece of code together with `Arrays.fill(bestX, 0.0);` will make `bestX` point to the shared `_tbest` array and fill it with zeros. Is this intentional ?

Comment (Pscheidl) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: Not crucial though ...

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java [hunk 3]
```diff
@@ -278,24 +305,28 @@ public static int rank(boolean transposed, double[][] archetypes_raw) {
    */
   private double[] gradientL(double[] x, double[] a) {
     // Prepate output row
-    double[] grad = new double[_ncolX];
-
+    //double[] grad = new double[_ncolX];
```
Comment (Pscheidl) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: remove the old line ?

**Comment by wendycwong** [comment]
Discussed with @pscheidl.  Getting rid of the  memory allocation may not gain us anything.  Let's do a test in Hadoop to find out if this thing helps with anything at all.  Please post timing information here before completing this PR.

**Comment by wendycwong** [comment]
I have completed a timing test on mr-0xd2, 4 cluster with 40G each.  Pavel is right, I do not have a lot of gain in speed.  However, this fix is more to enable users to run without running out of memory.  Previous implementation performs a lot of new double[] operation.  I generate 3 enum columns with various cat levels and then perform GLRM on the synthetic frame.  Refer to the JIRA for a detail timing result.  Here, I am just adding the highlights:


cat level | glrm runtime gain (%)
3000      | 1.848860781
2000      | 2.685828768
1000       | 3.34021701

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java (hunk 1 above)
Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: Done

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java (hunk 3 above)
Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmMojoModel.java]: Done

**Comment by wendycwong** [comment]
Small improvement in memory.  Not worth the change for the time being.

**Pull request closed without merging**